server/join.go

```diff
@@ -29,36 +30,118 @@ import (
+//        and it is not a member of cluster, join does MemberAdd, otherwise
+//        return an error.
+//
+// Etcd automatically re-join cluster if there is a data dir, so first we check
+// if there is data dir or not. With data dir, it returns an empty string(etcd
+// will get correct configurations from data dir.)
+//
+// Without data dir, we may have following cases:
+//
+//  1. a new pd joins to an existing cluster.
+//      join does: MemberAdd, MemberList, then generate initial-cluster.
+//
+//  2. a new pd joins itself
+//      join does: nothing.
+//
+//  3. an failed pd re-joins to previous cluster.
+//      join does: return an error(etcd reports: raft log corrupted, truncated,
+//                 or lost?)
+//
+//  4. a join self pd failed and it restarted with join while other peers try
+//     to connect to it.
+//      join does: nothing. (join can not detect whether it is in a cluster or
+//                 not, however, etcd will handle it safey, if there is no data
+//                 in cluster the restarted pd will join to cluster, otherwise
+//                 pd will shutdown as soon as other peers connect to it. etcd
+//                 reports: raft log corrupted, truncated, or lost?)
+//
+//  5. a deleted pd joins to previous cluster.
+//      join does: same as case1. (it is not in member list and there is no
+//                 data, so we can treat it as a new pd.)
+//
+// With data dir, special case:
+//
+//  6. a failed pd tries to join to previous cluster but it has been deleted
+//     during it's downtime.
+//      join does: return "" (etcd will connect to other peers and will find
+//                 itself has been removed)
+//
+//  7. a deleted pd joins to previous cluster
+//      join does: return "" (as etcd will read data dir and find itself has
+//                 been removed, so an empty string is fine.)
+func (cfg *Config) prepareJoinCluster() (string, string, error) {
+	initialCluster := ""
+	// cases with data
+	if wal.Exist(cfg.DataDir) {
+		return initialCluster, embed.ClusterStateFlagExisting, nil
+	}
+
+	// case 2, case 4
+	if cfg.Join == cfg.AdvertiseClientUrls {
+		initialCluster = fmt.Sprintf("%s=%s", cfg.Name, cfg.AdvertisePeerUrls)
+		return initialCluster, embed.ClusterStateFlagNew, nil
+	}
+
+	client, err := clientv3.New(genClientV3Config(cfg))
 	if err != nil {
-		return "", errors.Trace(err)
+		return "", "", errors.Trace(err)
 	}
 	defer client.Close()
 
-	addResp, err := client.MemberAdd(ctx, []string{cfg.AdvertisePeerUrls})
+	listResp, err := memberList(client)
 	if err != nil {
-		return "", errors.Trace(err)
+		return "", "", errors.Trace(err)
+	}
+
+	in := false
+	for _, m := range listResp.Members {
+		if m.Name == cfg.Name {
+			in = true
+		}
 	}
 
-	log.Infof("Member %16x added to cluster %16x\n", addResp.Member.ID, addResp.Header.ClusterId)
+	// case 3
+	if in {
+		return "", "", errors.New("missing raft log")
```

Comment:
the error `missing raft log` has no meaning. 
here seems that we want to join a duplicated pd to the cluster.